src/Bicep.Core/Workspaces/ProviderDescriptorBundle.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem.Providers;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+
+namespace Bicep.Core.Workspaces;
+
+public record ProviderDescriptorBundle(
+    IEnumerable<ResultWithDiagnostic<ResourceTypesProviderDescriptor>> ImplicitProviders,
+    ImmutableDictionary<ProviderDeclarationSyntax, ResultWithDiagnostic<ResourceTypesProviderDescriptor>> ExplicitProviderLookup);
+
+public class ProviderDescriptorBundleBuilder
+{
+    public ProviderDescriptorBundleBuilder(IEnumerable<ResultWithDiagnostic<ResourceTypesProviderDescriptor>> builtInProviders, ImmutableDictionary<ProviderDeclarationSyntax, ResultWithDiagnostic<ResourceTypesProviderDescriptor>> explicitProviderLookup)
+    {
+        foreach (var provider in builtInProviders)
+        {
+            AddImplicitProvider(provider);
+        }
+
+        foreach (var (providerDeclaration, providerDescriptor) in explicitProviderLookup)
+        {
+            AddOrUpdateExplicitProvider(providerDeclaration, providerDescriptor);
+        }
+    }
+
+    public ProviderDescriptorBundleBuilder() { }
+
+    private readonly HashSet<ResultWithDiagnostic<ResourceTypesProviderDescriptor>> implicitProviders = [];
```

Comment:
Any reason you've picked a mix of concurrent and non-conucurrent data structures?